### PR TITLE
Add verify golden test content functionalities

### PR DESCRIPTION
**Pull request by chloechien**
This introduces 2 dependencies (sergi/go-diff, fatih/color), both been adopted by Google opensource.
Screenshot 2023-12-10 at 3 58 16 PM

### CODE REVIEW

File: templates/commands/goldentest/verify.go [hunk 1]
```diff
@@ -60,9 +68,127 @@ func (c *VerifyCommand) Run(ctx context.Context, args []string) error {
...
+					failureText := red(fmt.Sprintf("-- [%s] expected, however missing", goldenFile))
+					err := fmt.Errorf(failureText)
+					tcErr = errors.Join(tcErr, err)
+					continue
+				} else {
+					return fmt.Errorf("failed to read (%s): %w", tempFile, err)
+				}
+			}
+
+			diffs := dmp.DiffMain(string(tempContent), string(goldenContent), false)
```
Comment (drevell) [inline templates/commands/goldentest/verify.go:132]: I'm not sure what the final parameter "false" means here, and the diffmatchpatch docs don't explain. Could you clarify?

File: templates/commands/goldentest/verify.go [hunk 2]
```diff
@@ -60,9 +68,127 @@ func (c *VerifyCommand) Run(ctx context.Context, args []string) error {
...
+			}
+
+			tempContent, err := os.ReadFile(tempFile)
+			if err != nil {
+				if errors.Is(err, os.ErrNotExist) {
+					failureText := red(fmt.Sprintf("-- [%s] expected, however missing", goldenFile))
+					err := fmt.Errorf(failureText)
+					tcErr = errors.Join(tcErr, err)
+					continue
+				} else {
```
Comment (drevell) [inline templates/commands/goldentest/verify.go:127]: Since the previous `if` block has a continue, the convention is to avoid the unnecessary "else" here. So instead of `else {foo}` it can just be `foo` . Same at line 116.

File: templates/commands/goldentest/verify.go [hunk 3]
```diff
@@ -60,9 +68,127 @@ func (c *VerifyCommand) Run(ctx context.Context, args []string) error {
...
+
+	if merr != nil {
+		return fmt.Errorf("golden test verification failure:\n %w", merr)
+	}
+
+	return nil
+}
+
+// addTestFile collects file paths generated in a golden test.
+func addTestFile(fileSet *map[string]struct{}, testDataDir string) error {
```
Comment (drevell) [inline templates/commands/goldentest/verify.go:164]: Perhaps addTestFile -> addTestFiles (plural)?

File: templates/commands/goldentest/verify.go [hunk 4]
```diff
@@ -60,9 +68,127 @@ func (c *VerifyCommand) Run(ctx context.Context, args []string) error {
...
+			return err
+		}
+		if err := addTestFile(&fileSet, tempDataDir); err != nil {
+			return err
+		}
+
+		dmp := diffmatchpatch.New()
+
+		var tcErr error
+		for relPath := range fileSet {
```
Comment (drevell) [inline templates/commands/goldentest/verify.go:104]: In Go, the iteration order of a map is undefined, which means the order of the file comparisons is undefined. We probably want to do something deterministic here, so the test output is the same every time. Perhaps alphabetical (which would implicitly also be depth-first if we include the directory name as part of the path)?

File: templates/commands/goldentest/verify.go [hunk 5]
```diff
@@ -60,9 +68,127 @@ func (c *VerifyCommand) Run(ctx context.Context, args []string) error {
...
+
+	for _, tc := range testCases {
+		goldenDataDir := filepath.Join(c.flags.Location, goldenTestDir, tc.TestName, testDataDir)
+		tempDataDir := filepath.Join(tempDir, goldenTestDir, tc.TestName, testDataDir)
+
+		fileSet := make(map[string]struct{})
+		if err := addTestFile(&fileSet, goldenDataDir); err != nil {
+			return err
+		}
+		if err := addTestFile(&fileSet, tempDataDir); err != nil {
```
Comment (drevell) [inline templates/commands/goldentest/verify.go:97]: Maybe if we used LoadDirWithoutMode we wouldn't need addTestFile?

https://github.com/abcxyz/abc/blob/eb69483b155ffb205abe88448047f27a6d1d90f4/templates/common/test_utils.go#L143

**Review by drevell — CHANGES_REQUESTED** [review]

File: templates/commands/goldentest/verify.go (hunk 5 above)
Comment (drevell) [reply templates/commands/goldentest/verify.go:97]: Oops, never mind, I just remembered that LoadDirWIthoutMode requires a *testing.T.

File: templates/commands/goldentest/verify.go [hunk 6]
```diff
@@ -60,9 +68,127 @@ func (c *VerifyCommand) Run(ctx context.Context, args []string) error {
...
 
-	if _, err := parseTestCases(ctx, c.flags.Location, c.flags.TestName); err != nil {
-		return fmt.Errorf("failed to parse golden test: %w", err)
+	testCases, err := parseTestCases(ctx, c.flags.Location, c.flags.TestName)
+	if err != nil {
+		return fmt.Errorf("failed to parse golden tests: %w", err)
+	}
+
+	// Create a temporary directory to render golden tests
+	tempDir, err := renderTestCases(testCases, c.flags.Location)
```
[the remark this answers is not in the document]
Comment (chloechien) [reply templates/commands/goldentest/verify.go:77]: I see, I was thinking of making golden test simple, but since we already have this patter we may consider adopt that in future.

File: templates/commands/goldentest/verify.go [hunk 7]
```diff
@@ -60,9 +68,127 @@ func (c *VerifyCommand) Run(ctx context.Context, args []string) error {
...
+	defer os.RemoveAll(tempDir)
+	if err != nil {
+		return fmt.Errorf("failed to render test cases: %w", err)
 	}
 
-	return fmt.Errorf("unimplemented")
+	var merr error
+	// Highlight error message color, given diff text might be hundreds lines long.
+	red := color.New(color.FgRed).SprintFunc()
```
[the remark this answers is not in the document]
Comment (chloechien) [reply templates/commands/goldentest/verify.go:85]: Good catch! I haven't thought about that, updated.

File: templates/commands/goldentest/verify.go [hunk 8]
```diff
@@ -60,9 +68,127 @@ func (c *VerifyCommand) Run(ctx context.Context, args []string) error {
...
+	green := color.New(color.FgGreen).SprintFunc()
+	resultReport := "\nTest Report:\n"
+
+	for _, tc := range testCases {
+		goldenDataDir := filepath.Join(c.flags.Location, goldenTestDir, tc.TestName, testDataDir)
```
[the remark this answers is not in the document]
Comment (chloechien) [reply templates/commands/goldentest/verify.go:90]: It seems that both a.. and ..a are interpreted as common folder name in filepath.join, and `..` is illegal to be a folder name. So we're fine here without extra trimming. But kindly lemme know if I'm wrong.

File: templates/commands/goldentest/verify.go [hunk 9]
```diff
@@ -60,9 +68,127 @@ func (c *VerifyCommand) Run(ctx context.Context, args []string) error {
...
+		tempDataDir := filepath.Join(tempDir, goldenTestDir, tc.TestName, testDataDir)
+
+		fileSet := make(map[string]struct{})
+		if err := addTestFile(&fileSet, goldenDataDir); err != nil {
```
[the remark this answers is not in the document]
Comment (chloechien) [reply templates/commands/goldentest/verify.go:94]: Got that! The magic of go

File: templates/commands/goldentest/verify.go (hunk 3 above)
Comment (chloechien) [reply templates/commands/goldentest/verify.go:164]: Done

File: templates/commands/goldentest/verify.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (drevell) [reply templates/commands/goldentest/verify.go:90]: The thing I'm worried about is the edge case where somebody has a TestName of `..`. As long as we have a test for that, or we're sure that will work OK, then I'm satisfied.

File: templates/commands/goldentest/verify.go (hunk 4 above)
Comment (chloechien) [reply templates/commands/goldentest/verify.go]: Good idea! Added.

File: templates/commands/goldentest/verify.go (hunk 2 above)
Comment (chloechien) [reply templates/commands/goldentest/verify.go]: Done.

File: templates/commands/goldentest/verify.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chloechien) [reply templates/commands/goldentest/verify.go:106]: I've tested with a.., ..a, a..a, both tests passed. And .. is not an illegal folder name so can't be retrieved as a testname here.

File: templates/commands/goldentest/verify.go [hunk 10]
```diff
@@ -60,9 +70,148 @@ func (c *VerifyCommand) Run(ctx context.Context, args []string) error {
...
+
+	// Highlight error message color, given diff text might be hundreds lines long.
+	// Only color the text when the result is to displayed at a terminal
+	var red, green func(a ...interface{}) string
+	useColor := c.Stdout() == os.Stdout && isatty.IsTerminal(os.Stdout.Fd())
+	if useColor {
+		red = color.New(color.FgRed).SprintFunc()
+		green = color.New(color.FgGreen).SprintFunc()
+	} else {
+		red = func(a ...interface{}) string {
```
Comment (drevell) [inline templates/commands/goldentest/verify.go:95]: Could this be simplified to

```
red = fmt.Sprint
green = fmt.Sprint
```

File: templates/commands/goldentest/verify.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (drevell) [reply templates/commands/goldentest/verify.go:106]: As long as it fails cleanly in the case where testName==".." that's all I care about.

**Review by drevell — APPROVED** [review]

File: templates/commands/goldentest/verify.go (hunk 10 above)
Comment (chloechien) [reply templates/commands/goldentest/verify.go]: Brilliant!

**Review by drevell — APPROVED** [review]

File: templates/commands/goldentest/verify.go (hunk 4 above)
Comment (drevell) [reply templates/commands/goldentest/verify.go]: Did this get fixed already? I think I'm still seeing the old version where the order is undefined

Comment (chloechien) [reply templates/commands/goldentest/verify.go]: Yes this is an older snapshots. Being fixed in the latest one.
